File: cousins_dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#APPROACH : DFS
#Time Complexity : O(n)
#Space Complexity : O(h) h-> height of tree
#Did your code run on leetcode: yes

class Solution:
    global x_depth,y_depth,x_parent,y_parent
    x_depth = 0
    y_depth = 0
    x_parent =  None
    y_parent = None
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        global x_depth,y_depth,x_parent,y_parent


        
        if root == None: return False
        
        self.dfs(root,x,y,0,None)
        return x_depth == y_depth and x_parent != y_parent
                
    def dfs(self, root: Optional[TreeNode], x: int, y: int,depth:int,parent :  Optional[TreeNode])->None:
        #base
        global x_depth,y_depth,x_parent,y_parent
        if(root == None) : return
        #logic
        if(root.val == x):
            x_depth = depth
            x_parent = parent
            
        if(root.val == y):
            y_depth = depth
            y_parent = parent
            
        self.dfs(root.left,x,y,depth+1,root)
        self.dfs(root.right,x,y,depth+1,root)


        

File: CousinsDFS.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
class Solution {
    TreeNode xParent;
    TreeNode yParent;
    int xLevel;
    int yLevel;
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null){
            return false;
        }
        if(x == y){
            return false;
        }
        
        return dfs(root, x, y, 0, null);
    }
    
    private boolean dfs(TreeNode root, int x, int y, int level, TreeNode parent){
        //base
        if(root == null){
            return false;
        }
        
        //logic
        // found x
        if(root.val == x){
            xParent = parent;
            xLevel = level;
        }
        
        // found y
        if(root.val == y){
            yParent = parent;
            yLevel = level;
        }
        
        if(xLevel == yLevel && xParent != yParent){
            return true;
        }
        
        return dfs(root.left, x, y, level+1, root) || dfs(root.right, x, y, level+1, root);
    }
}